sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -313,27 +315,24 @@ public FileBasedWriteOperation(FileBasedSink<T> sink) {
           sink.getBaseOutputFilenameProvider(), new TemporaryDirectoryBuilder()));
     }
 
-    private static class TemporaryDirectoryBuilder
+    @VisibleForTesting
+    static class TemporaryDirectoryBuilder
         implements SerializableFunction<String, String> {
       // The intent of the code is to have a consistent value of tempDirectory across
       // all workers, which wouldn't happen if now() was called inline.
       Instant now = Instant.now();
 
       @Override
       public String apply(String baseOutputFilename) {
-        try {
-          IOChannelFactory factory = IOChannelUtils.getFactory(baseOutputFilename);
-          Path baseOutputPath = factory.toPath(baseOutputFilename);
-          return baseOutputPath
-              .resolveSibling(
-                  "temp-beam-"
-                  + baseOutputPath.getFileName()
-                  + "-"
-                  + now.toString(DateTimeFormat.forPattern("yyyy-MM-DD_HH-mm-ss")))
-              .toString();
-        } catch (IOException e) {
-          throw new RuntimeException(e);
-        }
+        URI baseOutputUri = URI.create(baseOutputFilename).normalize();
```

Comment:
1.
My worries about adding this logic to PathUtils is that people might misuse it.
I think PathUtils should only contain common URI manipulation. And, they need to be simple and easy to understand to avoid being misused.
The logic here is not simple, and the client should own it: understand the URI manipulation, and verify the code with sufficient tests.

2. 
- non-URI paths needs to be converted before reaching to here.
- pathUri doesn't contains scheme, authority, query and fragment.
- self-relativize is to get the last component of pathUri.
"/home/output" resolve("") returns "/home/"
"/home/" relativize("/home/output") returns "output"
- line 332 is constructing the directory name. It is not path resolving (append /).

3.added test for now.